torch/_refs/__init__.py

```diff
@@ -0,0 +1,581 @@
+
+
+# TODO: implement ref with safe casting
+def _convert_dtype(*args, dtype: torch.dtype):
+    def _convert(x):
+        if isinstance(x, Tensor):
+            return prims.convert_element_type(x, dtype)
+        elif isinstance(x, Number):
+            typ = utils.dtype_to_type(dtype)
+            return typ(x)
+        return x
+
+    return tuple(map(lambda x: _convert(x), args))
+
+
+# TODO: handle tuples of tensors
+def _maybe_resize_out(out: Tensor, shape):
+    if out.numel() == 0:
+        return prims.resize(out, shape)
+
+    return out
+
+
+def _make_elementwise_binary_reference(prim: Callable, *, type_promotion) -> Callable:
+    def _ref(
+        a: Union[Tensor, Number], b: Union[Tensor, Number], out: Optional[Tensor] = None
+    ) -> Tensor:
+        assert isinstance(a, (Tensor, Number))
+        assert isinstance(b, (Tensor, Number))
+        assert out is None or isinstance(out, Tensor)
+
+        # Special-cases Number x Number case
+        if isinstance(a, Number) and isinstance(b, Number):
+            a, b = utils.wrap_scalars(a, b)
+
+        # Handles type promotion
+        computation_dtype, result_dtype = _elementwise_dtypes(
+            a, b, type_promotion=type_promotion
+        )
+        a, b = _convert_dtype(a, b, dtype=computation_dtype)
+
+        # Broadcasting
+        a, b = broadcast(a, b)
+
+        result = prim(a, b)
+
+        if type_promotion is not ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT:
+            (result,) = _convert_dtype(result, dtype=result_dtype)
+
+        if out is not None:
+            out = _maybe_resize_out(out, result.shape)
+            return copy_to(out, result, allow_cross_device=False)
+
+        return result
+
+    return _ref
+
+
+# Add is implemented specially because it has an alpha argument and
+#   is decomposed into multiple prims
+def add(
+    a: Union[Tensor, Number],
+    b: Union[Tensor, Number],
+    *,
+    alpha: Optional[Number] = None,
+    out: Optional[Tensor] = None
+):
+    """
+    Reference implementation of torch.add
+    """
+
+    # Type checks
+    assert isinstance(a, (Tensor, Number))
+    assert isinstance(b, (Tensor, Number))
+    assert out is None or isinstance(out, Tensor)
+    assert alpha is None or isinstance(alpha, Number)
+
+    # Special-cases Number x Number case
+    if isinstance(a, Number) and isinstance(b, Number):
+        a, b = utils.wrap_scalars(a, b)
+
+    computation_dtype, result_dtype = _elementwise_dtypes(
+        a, b, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.OP_MATH
+    )
+    a, b = _convert_dtype(a, b, dtype=computation_dtype)
+
+    a, b = broadcast(a, b)
+
+    if alpha is not None:
+        alpha_promotion_type = utils.dtype_to_type(computation_dtype)
+        assert utils.is_lesser_type(type(alpha), alpha_promotion_type) or (
+            computation_dtype is torch.bool and type(alpha) is int
+        )
+        b = prims.mul(b, alpha_promotion_type(alpha))
```

Comment:
Doesn't this rely on broadcasting, since `alpha` is a scalar? Or are primitives allowed to have multiple overloads and this is intended to call a `(Tensor, Scalar)` overload?

Reply:
That's a great question and yes, in our current thinking prims supports tensors and scalars.

Reply:
I am a bit confused. Why is scalar promotion not handled the same way as tensor promotion. I would expect it to be given that tensors and scalars can often be treated as one and the same (https://github.com/pytorch/pytorch/issues/75769#issuecomment-1099301958)

Reply:
I agree that would be natural but that's not how `torch.add` works